models/locked_resource.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2020 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package models
+
+import (
+	"fmt"
+
+	"code.gitea.io/gitea/modules/setting"
+)
+
+// LockedResource represents the locking key for a pessimistic
+// lock that can hold a counter
+type LockedResource struct {
+	LockType string `xorm:"pk VARCHAR(30)"`
+	LockKey  int64  `xorm:"pk"`
+	Counter  int64  `xorm:"NOT NULL DEFAULT 0"`
+
+	engine Engine `xorm:"-"`
```

Comment:
Could we still use engine as a parameter so that we can find the error obviously as before?

Reply:
I think this would have prevented errors instead (e.g. people calling on `x` would cause problems). I changed it, though.